### ISSUE TITLE
Use file-globs for sources

### ISSUE THREAD

**Issue by raveclassic** [esdoc/esdoc#72]
Hi!
First of all, I would like to thank you for this project - superawesome generator, the only one of its kind!

As for the issue - currently I'm trying to set up documentation generation for my project and I'm running into some troubles with `include/exclude` params in config.
First problem is that I cannot put my sources in one directory like it's suggested in manual (`src`), because of legacy support. All sources are in root directory. Ok, I can handle this by filling `exclude` array in config with `node_modules` etc.
Next problem is that the project is component-based and a single component directory contains lots of auxiliary `js` and `jsx` files, demo-pages, demo-scripts, etc. These files should not be documented, so I need to carefully set up `includes` array in config with heavy regexps just to filter out what I don't need in documentation.

So, my suggestion is to deprecate `sources`, `includes` and `excludes` and to use file-globs like it's done in `gulp.src`. It can be done either by specifying files as an array of glob-strings or by using array of plain file-paths.

Using globs in config:
```javascript
module.exports = {
  files: ['./util/**/*.js', './components/**/*.jsx', '!./components/**/*.test.jsx']
}
```

Using file paths as a result of https://github.com/isaacs/node-glob package:
```javascript
var glob = require('glob');
module.exports = {
  files: glob.sync(['./util/**/*.js', './components/**/*.jsx'], {
    ignore: ['./components/**/*.test.jsx']
  })
}
```

I think the second variant is more appropriate as it does not force you to use globs in library source and add an extra dependency - it's for the end-user.

What do you think? I could create a pull-request after approval of this suggestment.

**Comment by h13i32maru** [comment]
@raveclassic Hi! Thanks for this issue!
I understood your suggestion. But, for now, I do not have plan to support blob. Sorry.

I think that JavaScript de-facto standard directory structure has ``src`` or ``lib`` in root directory. So I think that ``config.source`` is enough. And If more complex directory structure, ``config.includes/config.excludes`` has powerful with regular-expression.

Your ``Using globs in config:`` is equal the following config.
```javascript
{
  source: "./",
  destination: "./doc",
  inclues: ["^util/.*\\.js$", "^components/.*\\.jsx$"],
  excludes: ["^components/.*\.test\\.jsx$"]
}
```

**Comment by raveclassic** [comment]
@h13i32maru Thanks for reply!
I get your point and agree that globs should not be supported by esdoc out of the box. What I mean is support of plain filepath arrays:
```javascript
var path = require('path');
module.exports = {
	cwd: process.cwd(), //this is current 'source' param
	files: [path.resolve('./components/first.jsx')/*absolute*/, './components/second.jsx'/*relative*/]
}
```
This is the easiest and the most flexible configuration and I don't need to mess with RegExp. 
The point is that `files` array can be set up using external tools such as node-glob or passed from upper environment (my case exactly, I don't have control of source files definition, I just have them as globs and can't convert easily to RegExp)
 
Combined with my previous pull-request about using commonjs in config files it's easy to use different tools, it's really flexible.

Surely it's up to you to decide but IMHO JS and node infrastructure are incredibly flexible tools and I suppose it would be great not to limit devs with forced project structure as I can never say for sure what structure I am forced to in another project.

Also, I can achieve what I need using `gulp.src` for creating fake vynil file-system and passing it to ESDoc, but that's not realy what I want. I don't want another tool such as gulp in my workflow until I really need it.

Thanks for support!

**Comment by h13i32maru** [comment]
@raveclassic Thanks for explanation!

What do you think the following changing?

- currently
  - ``config.source`` accepts directory path(string)
- changing
  - ``config.source`` accepts directory path(string) and **file/directory paths(string array)**

So, ESDoc will support the following both case.

```javascript
// case1: specify directory path
{
  source: "path/to/directory"
}
```

```javascript
// case2: specify file/directory paths
{
  source: ["path/to/directory", "path/to/file.js",,,]
}
```

**Comment by raveclassic** [comment]
@h13i32maru That would be an awesome feature for sure!
Regarding my suggestion about pull-request for this issue, if it's not urgent I could start working on this in two weeks.

Thanks!

**Comment by h13i32maru** [comment]
@raveclassic Please wait. Because I have some consideration for the changing.
So, I will consider whether to adopt the changes.

**Comment by ctrlplusb** [comment]
Any update on this?  I am also waiting for this. :innocent: 

Btw, thank you so much for this library!  It is amazing. :+1:

**Comment by agokjr** [comment]
This would be nice since I have server and client folders.

**Comment by h13i32maru** [comment]
I do not support this. Because multi root directory is very difficult and complex for ESDoc. Sorry.

**Comment by typhonrt** [comment]
I have created a comprehensive solution for multiple source path support with backward compatibility and an overview is available in #310.
